formation_mpc: # defined per generate_parameter_library
  control_horizon_in_s: {
    type: double,
    default_value: 0,
    validation: {
      gt<>: 0
    }
  }
  number_of_points: {
    type: int,
    default_value: 0,
    validation: {
      gt<>: 0
    }
  }
  dt: {
    type: double,
    default_value: 0.01,
    description: "Controller sampling time in seconds",
    validation: {
      gt<>: 0.0
    }
  }
  action_monitor_rate: {
    type: double,
    default_value: 100.0,
    description: "Monitor action state at defined frequency",
    validation: {
      gt<>: 0.0
    }
  }
  formation_dt: { # Not Implemented
    type: double,
    default_value: 0.1,
    description: "[NOT IMPLEMENTED] dt for formation mpc sync",
    validation: {
      gt<>: 0.0
    }
  }
  cartesian_axis: {
    type: bool_array,
    default_value: [true, true, true, true, true, true],
    description: "Cartesian axis to enable: [x, y, z, rx, ry, rz]. Default all enabled",
    validation: {
      fixed_size<>: 6
    }
  }
  clik:
    active: {
      type: bool,
      default_value: True,
      description: "Activate or deactivate clik"
    }
    gain: {
      type: double,
      default_value: 1.0,
      validation: {
        gt_eq<>: 0.0
      }
    }
  pid:
    P: {
      type: double_array,
      default_value: [],
      description: ""
    }
  leader_ns: {
    type: string,
    default_value: "sweepee_1",
    description: "Name of the formation leader"
  }
  is_omni: { # Not implemented
    type: bool,
    default_value: true,
    description: "[NOT IMPLEMENTED] is the robot omnidirectional?"
  }
  mobile_base_link: {
    type: string,
    default_value: "/base_footprint",
    description: "refered to the mobile base"
  }
  manipulator_base_link: {
    type: string,
    default_value: "/base_link",
    description: "manipulator base"
  }
  tool_frame: {
    type: string,
    default_value: "/tool0",
    description: "tool frame"
  }
  map_frame: {
    type: string,
    default_value: "map",
    description: "map/world reference frame"
  }
  mount_frame: {
    type: string,
    default_value: "base_link",
    description: "link to attach the fake_base"
  }
  robot_description: {
    type: string,
    default_value: "",
    description: "Robot description"
  }
  joints:
    base: {
      type: string_array,
      default_value: [],
      description: "Which joints belong to the mobile base"
    }
    arm: {
      type: string_array,
      default_value: [],
      description: "Which joints belong to the arm"
    }
  initial_positions: {
    type: double_array,
    default_value: [],
    description: "",
    validation: {
      not_empty<>: null
    }
  }
  command_joints: {
    type: string_array,
    default_value: [],
    description: "Name of the joints to control"
  }
  command_interface: {
    type: string_array,
    default_value: [],
    description: "Names of the control_interface to command",
  }
  state_interface: {
    type: string_array,
    default_value: [],
    description: "Names of the state_interface",
  }
  cmd_vel_topic: {
    type: string,
    default_value: "cmd_vel",
    description: "Name of the cmd_vel topic on which the base will read it cmd",
  }
  cmd_trj_topic: {
    type: string,
    default_value: "joint_trajectory_controller/joint_trajectory",
    description: "Topic of the joint_trajectory_controller",
  }
  cartesian_task_weight:
    active: {
      type: bool,
      default_value: false,
      description: "activate or deactivate scaling for cartesian task: ||J*qp-s*V||^2 + ||1-s||^2"
    }
    weight: {
      type: double,
      default_value: 1.0,
      description: "Scaling maximum value",
      validation: {
        lt_eq<>: 2.0
      }
    }
  scaling: {
    type: double,
    default_value: 1.0,
    description: "Scaling limit"
  }
  hierarchy: { # Not implemented
    type: int_array,
    default_value: [1,2,3],
    description: "hierarchy of the tasks (1 is higher). In order: minimize_acceleration, minimize_velocity, cartesian_trajectory",
    validation: {
      unique<>: [],
      fixed_size<>: 3,
      element_bounds<>: [1,3],
    }
  }
  trj_leader_topic: {
    type: string,
    default_value: "",
    description: "Topic though which the leader solution is shared with the followers"
  }
  plan_topic: { # Not implemented (and not useful)
    type: string,
    default_value: "/plan",
    description: "Topic with the global plan for the formation",
  }
  base_max_vel: {
    type: double_array,
    default_value: [0.0,0.0,0.0,0.0,0.0,0.0],
    description: "Maximum velocity of the payload",
    validation: {
      fixed_size<>: [6],
      lower_element_bounds<>: 0.0
    }
  }
  fake_base:
    package: {
      type: string,
      default_value: "",
      description: "Package inside which there is the xacro"
    }
    path: {
      type: string,
      default_value: "",
      description: "Relative path to xacro from <package>/share/"
    }
